caffe2/operators/experimental/c10/schemas/layer_norm.h

```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include "caffe2/core/tensor.h"
+#include <c10/util/Array.h>
+
+namespace caffe2 {
+namespace ops {
+
+struct LayerNorm final {
+  static constexpr const char* name = "LayerNorm";
+
+  struct Cache final {
+    Tensor scale = Tensor{CPU};
+    Tensor bias = Tensor{CPU};
+  };
+
+  using Signature = void(
+      const Tensor& input,
+      Tensor* output,
+      Tensor* output_mean,
+      Tensor* output_stddev,
+      int axis,
+      float epsilon,
+      Cache* cache,
+      BaseContext* context);
```

Comment:
passing the context here is unfortunate. You seem to need it for getting current stream id. This part is not unified yet (cc @ezyang) - pytorch and c2 would create different streams.

I guess, you can skip CUDA for now. For C2 implementation you can add a static method here (https://github.com/pytorch/pytorch/blob/master/caffe2/core/context_gpu.h#L222) that gives cuda_stream(CaffeCudaGetDevice()). It'd add a small overhead of extra GetDevice but would eliminate the need of passing context.

It would be not properly safe to call from PyTorch side though as the stream id would be different (until Ed's stuff lands). You can wrap it in explicit CUDAContext instance there effectively making the call blocking for now (since destructor invokes FinishDeviceComputation).

Reply:
When calling from caffe2, I'm getting the context from the Operator calling it, so that should be routed correctly and also work for CUDA. Calling CUDA from PyTorch won't work for the reasons you mentioned.